### PR TITLE
Feature/GV-98 Unreal GDK multiple engine version CI

### PR DESCRIPTION
**Pull request by ImprobableVlad**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
In the near future we would like to start officialy supporting multiple Unreal Engine versions with the GDK for Unreal. Therefore evaluating multiple engine versions in CI is necessary. This PR includes the necessary code for enabling multi-version building and testing on Buildkite.

##### Use
All engine versions that are listed in the `unreal-engine.version` file are automatically tested. Previously listing only a single version was supported.

##### Changes
The engine-building steps and their dependencies that were previously included in `.buildkite/premerge.steps.yaml` have been moved to their own `ci/gdk_buil_template.steps.yaml`. In their place, `.buildkite/premerge.steps.yaml` has now been extended with an extra step that dynamically adds steps for each engine version. In essence, the `ci/gdk_buil_template.steps.yaml` file serves as a template that is filled out and reuploaded for every listed engine version.

Finally, `check-version-file.sh` has been extended to support multiple engines. Protected branches now cannot accept a merge where _any_ listed version is unpinned.

#### Release note
This change is not customer facing. No release note has been added.

#### Tests
Adding the 4.23 engine version to `unreal-engine.version` produced the expected build results with new steps being dynamically added: https://buildkite.com/improbable/unrealgdk-premerge/builds/6600 . The 4.23 builds fail due to using a GDK version that does not support 4.23 - which is indeed the expected behavior (what matters is that they ran).

STRONGLY SUGGESTED: How can this be verified by QA?
Branch off this branch and modify `unreal-engine.version` to contain no or additional (possibly faulty) engine versions. Verify that the expected versions are (not) being built on the build's BK page.

#### Documentation
In-code documentation is included.

#### Primary reviewers
@mironec @joshuahuburn @acheronehj

### CODE REVIEW

File: ci/check-version-file.sh [hunk 1]
```diff
@@ -24,16 +35,18 @@ do
...
+    IFS=$'\n'
+    for engine_version in $(cat < ci/unreal-engine.version); do
+        echo "Found engine version $engine_version"
 
-    if [[ $engine_version == HEAD* ]]; then # version starts with "HEAD"
-        error_msg="The merge target branch does not allow a floating (HEAD) engine version. Use a pinned version. (Of the form UnrealEngine-{commit hash})"
-        
-        echo $error_msg | buildkite-agent annotate --context "check-version-file" --style error
+        if [[ $engine_version == HEAD* ]]; then # version starts with "HEAD"
+            error_msg="The merge target branch does not allow a floating (HEAD) engine versions. Use pinned versions. (Of the form UnrealEngine-{commit hash})"
```
Comment (mironec) [inline ci/check-version-file.sh]: Nit: 
```suggestion
            error_msg="The merge target branch does not allow floating (HEAD) engine versions. Use pinned versions. (Of the form UnrealEngine-{commit hash})"
```

File: ci/generate_and_upload_build_steps.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+set -euo pipefail
+
+# This script generates steps for each engine version listed in unreal-engine.version, and adds those to generated_base.steps.yaml
+# The steps are based on the template in generated_steps.steps.yaml
+
+if [ -z "${ENGINE_VERSION}" ]; then 
+    echo "Generating build steps for each engine version listed in unreal-engine.version"  
+    IFS=$'\n'
+    for commit_hash in $(cat < ci/unreal-engine.version); do
+        sed "s/INSERT_ENGINE_COMMIT_HASH/$commit_hash/g" ci/gdk_build_template.steps.yaml | buildkite-agent pipeline upload
```
Comment (mironec) [inline ci/generate_and_upload_build_steps.sh]: I think I prefer something else than `INSERT_ENGINE_COMMIT_HASH` - maybe `ENGINE_COMMIT_HASH_PLACEHOLDER`?

File: ci/unreal-engine.version [hunk 3]
```diff
@@ -1 +1,2 @@
 HEAD 4.22-SpatialOSUnrealGDK
+HEAD 4.23-SpatialOSUnrealGDK-merge
```
Comment (mironec) [inline ci/unreal-engine.version]: I think remove this for now - just CI against 4.22-SpatialOSUnrealGDK, and we will do the CI against two versions in the 4.23 branch

File: ci/generate_and_upload_build_steps.sh (hunk 2 above)
Comment (ImprobableVlad) [reply ci/generate_and_upload_build_steps.sh]: Yeah, that does sound better to me as well. I've fixed it now.

File: ci/generate_and_upload_build_steps.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+set -euo pipefail
+
+# This script generates steps for each engine version listed in unreal-engine.version, and adds those to generated_base.steps.yaml
+# The steps are based on the template in generated_steps.steps.yaml
+
+if [ -z "${ENGINE_VERSION}" ]; then 
+    echo "Generating build steps for each engine version listed in unreal-engine.version"  
+    IFS=$'\n'
```
Comment (joshuahuburn) [inline ci/generate_and_upload_build_steps.sh]: For my own knowledge, could you explain how this for loop knows how to delimit via the newline

Comment (ImprobableVlad) [reply ci/generate_and_upload_build_steps.sh]: Sure thing. In bash `IFS` is a special "internal field separator" variable. That variable is used in, among other places, the `for` construct to determine what to use as delimiters. In our case, `IFS` is set to a newline character on line 9 (IFS=$'\n').

**Pull request merged**